### PR TITLE
Drivers

### PR DESCRIPTION
**Pull request by paulpatience**

### CODE REVIEW

**Comment by paulpatience** [comment]
I've changed the directory structure as you requested.

What should the API look like?

**Comment by pavel-kirienko** [comment]
I suggest this:

* Put the driver into just one source/header file pair, named `socketcan.c`/`socketcan.h`, to simplify inclusion.
* Prefix all driver definitions with `socketcan` or `SOCKETCAN`, as appropriate. I don't think we need to use the Libcanard prefix here.
* The driver should define at least the following functions (documented in the header):
  * Initialization, e.g. `int socketcanInit(SocketCANInstance* ins, const char* can_iface_name)`
  * Function that closes the socket and deinitializes everything, e.g. `int socketcanClose(SocketCANInstance* ins)`.
  * Transmission, e.g. `int socketcanTransmit(SocketCANInstance* ins, const CanardCANFrame* frame)`
  * Reception with timeout, e.g. `int socketcanReceive(SocketCANInstance* ins, CanardCANFrame* out_frame, unsigned timeout_msec)`
  * We will also need to define a way for the application to perform IO multiplexing on the CAN socket. For this, we'll need a function that returns the file descriptor of the CAN socket, e.g. `int socketcanGetSocketFileDescriptor(SocketCANInstance* ins)`, or something alike.

I am open to any suggestions.

**Comment by paulpatience** [comment]
I've made an initial attempt at the socketcan driver.

A few questions I have:
* Should we use ppoll() or poll()? I saw libuavcan uses ppoll() but doesn't use the sigmask argument. The only difference is then in the timeout argument.
* Should we allow the user to specify an infinite timeout, for example by specifying a timeout of -1 to poll() or NULL to ppoll()? In this case the socketcan timeout argument may need to be signed.
* Should there also be a timeout argument in socketcanTransmit()? It would be symmetric with socketcanReceive().
* Is the int return value in socketcanClose() so that the user can tell if close() failed or are there other failures you think are important? libuavcan doesn't check the return value of close(). close(2) says to use the return value only for diagnostics. I suppose in the future the driver could have additional deinitialization steps that could fail.

**Comment by pavel-kirienko** [comment]
1. Libuavcan uses `ppoll()` only because it needs timeouts with higher granularity, so you can use `poll()` here.

2. Yes, okay, signed, makes sense.

3. Right, I agree.

4. The main reason is consistency - I think all or almost all other API calls in the library return int, so I figured it would make sense to follow this pattern here as well. Also, as you said, the deinitialization function may have additional logic in the future which may make use of return codes.

**Comment by paulpatience** [comment]
I've added a nuttxcan driver.

A few more questions:
* The equivalent to socketcanGetSocketFileDescriptor() is nuttxcanGetDeviceFileDescriptor(). Would it be better to name them socketcanGetFileDescriptor() and nuttxcanGetFileDescriptor() to make them more uniform?
* The transmit and receive functions return -1 when poll() times out. Is this what we want, or should we indicate that a timeout occurred?
* Is there a particular reason the functions return 1 on success instead of 0?
* Will the library one day return different values for different failures or successes?

**Comment by pavel-kirienko** [comment]
1. If you feel that it will make sense, sure, although I don't think uniformity here is necessary, since SocketCAN and NuttX's CAN driver are very different.

2. Libuavcan uses the following logic:
  * On error, functions return negated error codes
  * If nothing happened (e.g. send/receive function timed out), returned value is zero
  * If the frame has been transferred successfully, returned value is 1 (because one frame)

3. See above.

4. Zero will always be used for success; however, error codes will be changed; see https://github.com/UAVCAN/libcanard/issues/8

**Comment by paulpatience** [comment]
Ok, I've left socketcanGetSocketFileDescriptor() and nuttxcanGetDeviceFileDescriptor() as is and made the transmit and receive functions return 0 on timeout.

**Comment by pavel-kirienko** [comment]
Cool, let me know when it's ready for review.

**Comment by paulpatience** [comment]
It's ready for review.

**Comment by paulpatience** [comment]
Hm, I just rerealized that NuttX currently has no support for polling CAN devices. I will add that support maybe tomorrow, then.

**Comment by pavel-kirienko** [comment]
Hi Paul,

I think there was a minor misunderstanding. My proposition concerning error reporting was to return non-negative if the function succeeded, and negative in case of failure. Functions that cannot return a meaningful positive value should return zero on success. This applies, for example, to `nuttxcanInit()`, which should return 0 on success; whereas functions that can return meaningful information, such as `nuttxcanTransmit()`, should return 0 on timeout (which is also considered success) and 1 on successful transmission.

I will add a few more notes in a second. Shout out to @masierra - we're slightly altering your testing app, please have a glance.

File: drivers/nuttxcan/nuttxcan.c [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    out_ins->fd = fd;
+    return 1;
+}
+
+/**
+ * Deinitializes the NuttXCAN instance.
+ */
+int nuttxcanClose(NuttXCANInstance* ins)
+{
+    const int close_result = close(ins->fd);
```
Comment (pavel-kirienko) [inline drivers/nuttxcan/nuttxcan.c]: Considering my last comment on return codes, you could just `return close_result;`. I also suggest to invalidate the `fd` after closing, e.g. make it -1.

File: drivers/nuttxcan/nuttxcan.c [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+#include "nuttxcan.h"
+#include <fcntl.h>
+#include <poll.h>
+#include <string.h>
+#include <unistd.h>
+#include <nuttx/can.h>
+
+/**
+ * Initializes the NuttXCAN instance.
+ */
```
Comment (pavel-kirienko) [inline drivers/nuttxcan/nuttxcan.c]: I strongly suggest to move all documentation comments into the header file. The main library partially deviates from this convention, but it should be fixed someday.

File: drivers/nuttxcan/nuttxcan.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2016 UAVCAN Team
+ *
+ * Distributed under the MIT License, available in the file LICENSE.
+ *
+ */
+
+#include "canard.h"
```
Comment (pavel-kirienko) [inline drivers/nuttxcan/nuttxcan.h]: This is not a local inclusion, so it should be changed to `#include <canard.h>`

**Comment by pavel-kirienko** [comment]
Do you think `nuttxcan` is an acceptable name? I looked at the NuttX documentation and noticed that the CAN driver does not have a name of its own, so I suggest to rename the directory and the files to `nuttx`. SocketCAN is alright since it's a well-known name.

**Comment by paulpatience** [comment]
I've made the changes you requested. The reason I chose the name nuttxcan is that if we suppose choosing nuttx as the driver name is fine, some other library could choose to name a driver that interfaces with NuttX the same way (instead of calling it nuttxdev where dev is some device). It's less likely for users of libcanard to be simultaneously using another library that uses the NuttX CAN driver and calls it nuttxcan.

I suppose it depends on the names you want for the other drivers, e.g., stm32can or just stm32? In the same way, stm32 could clash with other libraries.

While we're on the subject of return values, I noticed some functions in canard.c return a uint64_t and always return 0 (e.g., canardReleaseRxTransferPayload()). Is this correct? If there is a chance they might fail in the future they should return an int, otherwise we could make them void.

**Comment by pavel-kirienko** [comment]
Re return values - `canardReleaseRxTransferPayload()` is supposed to return nothing. I just fixed the mistake in the design document; it should be fixed in the code too (if you fixed it in your PR we'd gladly accept it).

As for naming - your concerns about naming conflicts are valid. I suggest to rename the driver to `nuttx` and prefix declarations with `canard`, e.g. `canardNuttXTransmit()`. Other drivers that are susceptible to name clashing should name their entities likewise, e.g. `canardSTM32Receive()`. Does it make sense? I don't like the current naming style (`nuttxcan`) because it doesn't reflect the semantics. All of this does not apply to SocketCAN though - it should be conflict proof enough as it is now.

**Comment by paulpatience** [comment]
I renamed `NuttXInstance` to `CanardNuttXInstance` and prefixed the functions with `canard`. Will the headers still be nuttx.h, stm32.h, etc.? There might be conflicts, for example with NuttX's stm32.h, which we sometimes need to include to get access to some definitions. The user could add the libcanard directory as an include path and then include `<libcanard/canard.h>` and `<libcanard/drivers/nuttx.h>`, but then the way we've included `canard.h` in the drivers themselves wouldn't work. Should we move all the includes into a `canard` directory?

Fixed `canardReleaseRxTransferPayload()`.

**Comment by pavel-kirienko** [comment]
Great. I suggest to prefix headers with `canard_`, that should take care of
possible conflicts.

**Comment by paulpatience** [comment]
Should we prefix the `socketcan` header as well, or just the others?

**Comment by pavel-kirienko** [comment]
Whatever you find appropriate.

**Comment by paulpatience** [comment]
Ok, I've renamed `nuttx.h` to `canard_nuttx.h` and left `socketcan.h` as is.

**Comment by pavel-kirienko** [comment]
I don't have anything to add. Is this ready for merge?

**Comment by paulpatience** [comment]
Yes.

**Comment by pavel-kirienko** [comment]
Thanks Paul!

**Pull request merged**